src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -802,6 +803,30 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded, Immu
                 }
             }
 
+            private static DocumentId CreateStableSourceGeneratedDocumentId(ProjectId projectId, ISourceGenerator generator, string hintName)
+            {
+                // We want the DocumentId generated for a generated output to be stable between Compilations; this is so features that track
+                // a document by DocumentId can find it after some change has happened that requires generators to run again.
+                // To achieve this we'll just do a crytographic hash of the generator name and hint name; the choice of a cryptographic hash
+                // as opposed to a more generic string hash is we actually want to ensure we don't have collisions.
+                var generatorName = generator.GetType().FullName;
+                Contract.ThrowIfNull(generatorName);
+
+                // Combine the strings together; we'll use Encoding.Unicode since that'll match the underlying format; this can be made much
+                // faster once we're on .NET Core since we could directly treat the strings as ReadOnlySpan<byte>.
+                using var _ = ArrayBuilder<byte>.GetInstance(capacity: (generatorName.Length + hintName.Length + 1) * 2, out var hashInput);
+                hashInput.AddRange(Encoding.Unicode.GetBytes(generatorName));
+                hashInput.AddRange(0, 0); // null to separate the generator name and hint name
+                hashInput.AddRange(Encoding.Unicode.GetBytes(hintName));
+
+                // The particular choice of crypto algorithm here is arbitrary and can be always changed as necessary.
+                var hash = System.Security.Cryptography.SHA256.Create().ComputeHash(hashInput.ToArray());
+                Array.Resize(ref hash, 16);
```

Comment:
It says above we're using a crypto hash to avoid collisions, but if we're only using the first 16 bits, then that isn't guaranteed. Should the comment above be 'to reduce the likelihood of collisions?'

Reply:
16 bytes, not bits. I tried (and failed) to find the similar aproach in the compiler where we compute the MVID in deterministic build cases -- do you know where that is?